File: api-java/src/main/java/io/github/regextester/api/service/TelemetryService.java
package io.github.regextester.api.service;

import com.azure.cosmos.CosmosClient;
import com.azure.cosmos.CosmosClientBuilder;
import com.azure.cosmos.CosmosContainer;
import com.azure.identity.DefaultAzureCredentialBuilder;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Cosmos DB telemetry for api-java.
 *
 * <p>Mirrors the .NET, Node.js and Python implementations: Entra ID authentication via
 * {@code DefaultAzureCredential} with no account key anywhere, silently disabled when
 * {@code COSMOS_ENDPOINT} is empty, {@code /timestamp} partition key, the standardized
 * 12-field document, and strictly fire-and-forget so a Cosmos outage can never affect the
 * {@code POST /api/regex} response. api-dotnet once awaited its write, so an outage returned HTTP
 * 500 to users; do not reintroduce that here.
 *
 * <p>No client IP is ever collected — {@code host} is the Host header.
 */
public class TelemetryService {

    private static final Logger log = LoggerFactory.getLogger(TelemetryService.class);

    private final String endpoint;
    private final String databaseName;
    private final String containerName;

    /**
     * Single daemon thread: telemetry must never keep the JVM alive, and serializing writes bounds
     * the resource cost of a Cosmos slowdown to one thread instead of the request pool.
     */
    private final ExecutorService executor = Executors.newSingleThreadExecutor(runnable -> {
        Thread thread = new Thread(runnable, "telemetry");
        thread.setDaemon(true);
        return thread;
    });

    /**
     * Upper bound on Cosmos initialization. Init blocks the startup thread, so an unreachable
     * endpoint would otherwise hang startup and turn a telemetry outage into a total outage.
     */
    private static final long INIT_TIMEOUT_MS = 10_000;

    private volatile CosmosClient client;
    private volatile CosmosContainer container;

    public TelemetryService(String endpoint, String databaseName, String containerName) {
        this.endpoint = endpoint;
        this.databaseName = databaseName;
        this.containerName = containerName;
    }

    /**
     * Initialize the Cosmos client during startup. No-op when the endpoint is empty.
     *
     * <p>Blocks the startup thread until the client is ready, so the very first {@code
     * POST /api/regex} after a restart is recorded rather than silently dropped — App Service
     * restarts instances routinely, so a warm-up window is continuous data loss, not an edge case.
     *
     * <p>The work runs on the telemetry executor purely so {@link Future#get(long, TimeUnit)} can
     * bound it: the Java Cosmos SDK offers no per-call timeout for these operations. On timeout the
     * abandoned attempt is left to finish on its own; if it eventually succeeds it legitimately
     * leaves a usable client behind. Any failure is logged at warning level and leaves telemetry
     * disabled — a bad or unreachable endpoint, a missing role assignment or an unavailable
     * credential must never prevent the app from starting.
     */
    public void init() {
        if (endpoint == null || endpoint.isBlank()) {
            return;
        }
        Future<?> initTask = executor.submit(() -> {
            // Entra ID, never an account key: DefaultAzureCredential resolves the App Service
            // managed identity in Azure and the developer's az login session locally. A rotated
            // key silently disabled telemetry for five weeks in 2026-07; there is now no key.
            CosmosClient built = new CosmosClientBuilder()
                    .endpoint(endpoint)
                    .credential(new DefaultAzureCredentialBuilder().build())
                    .buildClient();
            CosmosContainer cont = built.getDatabase(databaseName).getContainer(containerName);

            // getDatabase/getContainer only build client-side handles, so without this read the
            // first token acquisition — and any 403 from a missing role assignment — would be
            // deferred to the first write and lost in its catch. One metadata round trip here
            // proves the identity can reach the container, and is covered by the readMetadata
            // action of Cosmos DB Built-in Data Contributor. It replaces the two
            // createIfNotExists calls, which that role deliberately cannot perform: creating a
            // database or container is a control-plane operation. The container is provisioned by
            // DEPLOYMENT.md §2 and must already exist, partitioned on /timestamp.
            cont.read();

            this.container = cont;
            this.client = built;
        });

        try {
            initTask.get(INIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
        } catch (TimeoutException e) {
            log.warn("Cosmos DB telemetry initialization timed out after {} ms; telemetry is disabled.",
                    INIT_TIMEOUT_MS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            log.warn("Cosmos DB telemetry initialization was interrupted; telemetry is disabled.");
        } catch (Exception e) {
            log.warn("Cosmos DB telemetry initialization failed; telemetry is disabled.", e);
        }
    }

    /**
     * Build the standardized 12-field telemetry document. Pure function, no I/O — kept separate from
     * {@link #send} so the shape can be verified without a live Cosmos account.
     *
     * <p>A {@link LinkedHashMap} preserves field order in the stored document, matching the other
     * three engines.
     */
    public Map<String, Object> buildDocument(
            String host,
            String userAgent,
            String pattern,
            String text,
            String replace,
            int options,
            long durationMs,
            int matchCount,
            String error) {

        Map<String, Object> document = new LinkedHashMap<>();
        document.put("id", UUID.randomUUID().toString());
        document.put("engineKey", CapabilitiesService.ENGINE_KEY);
        document.put("timestamp", Instant.now().toString());
        document.put("host", host == null ? "" : host);
        document.put("userAgent", userAgent == null ? "" : userAgent);
        document.put("pattern", pattern);
        document.put("text", text);
        document.put("replace", replace);
        document.put("options", options);
        document.put("durationMs", durationMs);
        document.put("matchCount", matchCount);
        document.put("error", error);
        return document;
    }

    /**
     * Write a pre-built document to Cosmos.
     *
     * <p>Fire-and-forget: the caller never blocks on it, and every exception is swallowed here. The
     * SDK reads the partition key from the document body, so — like api-nodejs and api-python, and
     * unlike api-dotnet — no explicit partition key value is passed.
     */
    public void send(Map<String, Object> document) {
        CosmosContainer target = container;
        if (target == null) {
            return;
        }
        executor.execute(() -> {
            try {
                target.createItem(document);
            } catch (RuntimeException e) {
                log.warn("Telemetry write to Cosmos DB failed: {}", e.getMessage());
            }
        });
    }

    public void shutdown() {
        executor.shutdownNow();
        CosmosClient open = client;
        if (open != null) {
            open.close();
        }
    }
}

File: api-dotnet/Services/TelemetryService.cs
using Azure.Identity;
using Microsoft.AspNetCore.Http;
using Microsoft.Azure.Cosmos;
using Microsoft.Extensions.Logging;
using RegExTester.Api.DotNet.Models;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace RegExTester.Api.DotNet.Services;


public interface ITelemetryService
{
    /// <summary>
    /// Records one telemetry document for a completed <c>POST /api/regex</c> request.
    /// Fire-and-forget: dispatches the Cosmos write on a background task and returns
    /// immediately, so a Cosmos outage can never delay or fail the HTTP response.
    /// </summary>
    void RecordTelemetry(HttpRequest request, Input model, TimeSpan duration, int matchCount, string error);
}

public class TelemetryService : ITelemetryService, IDisposable
{
    public static ItemRequestOptions ItemRequestOptions = new ItemRequestOptions { EnableContentResponseOnWrite = false };

    /// <summary>
    /// Upper bound on Cosmos initialization. Init runs on the startup path, so an unreachable
    /// endpoint would otherwise hang startup and turn a telemetry outage into a total outage.
    /// </summary>
    private static readonly TimeSpan InitTimeout = TimeSpan.FromSeconds(10);

    public static CosmosClient CosmosClient { get; private set; }
    public static Database CosmosDatabase { get; private set; }
    public static Container CosmosContainer { get; private set; }

    private readonly ILogger<TelemetryService> _logger;

    public TelemetryService(string cosmosEndpoint, string database, string container, ILogger<TelemetryService> logger)
    {
        _logger = logger;
        InitCosmos(cosmosEndpoint, database, container);
    }

    public void RecordTelemetry(HttpRequest request, Input model, TimeSpan duration, int matchCount, string error)
    {
        if (CosmosClient is null || CosmosContainer is null)
            return;

        var item = new
        {
            id = Guid.NewGuid().ToString(),
            engineKey = RegExTesterOptionsRegistry.EngineKey,
            timestamp = DateTime.UtcNow.ToString("o"),
            host = request.Host.Value,
            userAgent = request.Headers["User-Agent"].ToString(),
            pattern = model.Pattern,
            text = model.Text,
            replace = model.Replace,
            options = (int)model.Options,
            durationMs = (int)duration.TotalMilliseconds,
            matchCount,
            error
        };

        // Fire-and-forget: never awaited by the caller, and CancellationToken.None is used
        // (not the request's token, which may already be cancelled by the time the response
        // has been sent). Every exception is caught inside this task so none can escape as an
        // unobserved task exception, and telemetry can never affect a response already sent.
        _ = Task.Run(async () =>
        {
            try
            {
                // The partition key value MUST match the provisioned container's partition key
                // path (/timestamp). Passing engineKey here would make every write fail with
                // PartitionKeyMismatch, and because this catch swallows everything that failure
                // would be completely silent.
                await CosmosContainer.CreateItemAsync(item, new PartitionKey(item.timestamp), ItemRequestOptions, CancellationToken.None);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Telemetry write to Cosmos DB failed.");
            }
        });
    }

    private void InitCosmos(string cosmosEndpoint, string database, string container)
    {
        if (string.IsNullOrEmpty(cosmosEndpoint) || CosmosClient is not null)
            return;

        // Bounded out here rather than with a CancellationToken: the Cosmos SDK does not honour one
        // promptly against an unreachable endpoint (measured at ~37s against a blackholed address,
        // versus the 10s asked for). A connection that completes after the bound still publishes
        // its client — it is perfectly usable, it just missed the startup window.
        var init = Task.Run(() => ConnectAsync(cosmosEndpoint, database, container));

        if (!init.Wait(InitTimeout))
        {
            _logger?.LogWarning(
                "Cosmos DB telemetry initialization exceeded {TimeoutMs} ms; starting without it.",
                InitTimeout.TotalMilliseconds);
        }
    }

    private async Task ConnectAsync(string cosmosEndpoint, string database, string container)
    {
        CosmosClient client = null;
        try
        {
            // Entra ID, never an account key: DefaultAzureCredential resolves the App Service
            // managed identity in Azure and the developer's az login session locally. A rotated
            // key silently disabled telemetry for five weeks in 2026-07; there is now no key.
            client = new CosmosClient(cosmosEndpoint, new DefaultAzureCredential());
            var db = client.GetDatabase(database);
            var cont = db.GetContainer(container);

            // GetDatabase/GetContainer only build client-side handles, so without this read the
            // first token acquisition — and any 403 from a missing role assignment — would be
            // deferred to the first write and lost in its catch. One metadata round trip here
            // proves the identity can reach the container, and is covered by the readMetadata
            // action of Cosmos DB Built-in Data Contributor. It replaces the two
            // CreateIfNotExists calls, which that role deliberately cannot perform: creating a
            // database or container is a control-plane operation. The container is provisioned
            // by DEPLOYMENT.md §2 and must already exist, partitioned on /timestamp.
            await cont.ReadContainerAsync();

            // Published only once the round trip succeeded, so a partially initialized client is
            // never visible to RecordTelemetry.
            CosmosClient = client;
            CosmosDatabase = db;
            CosmosContainer = cont;
        }
        catch (Exception ex)
        {
            // Telemetry is non-essential: a bad, unreachable or slow endpoint, a missing role
            // assignment or an unavailable credential must never prevent the app from starting.
            // Log and leave telemetry disabled. Caught inside the task so a post-timeout failure
            // never surfaces as an unobserved task exception.
            _logger?.LogWarning(ex, "Cosmos DB telemetry initialization failed; telemetry is disabled.");
            client?.Dispose();
        }
    }

    public void Dispose()
    {
        CosmosContainer = null;
        CosmosDatabase = null;
        CosmosClient?.Dispose();
    }
}
